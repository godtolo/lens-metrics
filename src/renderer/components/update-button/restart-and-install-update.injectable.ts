import { getInjectable } from "@ogre-tools/injectable";
import restartAndInstallUpdateChannel from "../../../common/application-update/update-warning-level/restart-and-install-update-channel.injectable";
import messageToChannelInjectable from "../../utils/channel/message-to-channel.injectable";

const restartAndInstallUpdateInjectable = getInjectable({
  id: "restart-and-install-update",

  instantiate: (di) => {
    const messageToChannel = di.inject(messageToChannelInjectable);
    const channel = di.inject(restartAndInstallUpdateChannel);

    return () => {
      messageToChannel(channel);
    };
  },
});

export default restartAndInstallUpdateInjectable;