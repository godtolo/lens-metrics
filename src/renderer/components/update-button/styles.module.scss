/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

.updateButton {
  --active-color-hsl: 122deg 39% 49%;
  --active-color: hsl(var(--active-color-hsl));

  border: 1px solid var(--active-color);
  border-radius: 4px;
  background: hsl(var(--active-color-hsl) / 15%);
  color: var(--active-color);
  display: flex;
  align-items: center;
  padding: 4px 8px;
  gap: 8px;
  cursor: default;
  transition: background-color 0.1s;

  &:hover {
    background: hsl(var(--active-color-hsl) / 5%);
  }

  &:focus-visible {
    box-shadow: 0 0 0 2px var(--blue);
    border-color: transparent;
  }
}

.warningHigh {
  --active-color-hsl: 0deg 76% 62%;
}

.warningMedium {
  --active-color-hsl: 36deg 100% 50%;
}

:global(.theme-light) .warningMedium {
  // Making medium color a bit darker for light theme
  --active-color-hsl: 36deg 100% 46%;
}
