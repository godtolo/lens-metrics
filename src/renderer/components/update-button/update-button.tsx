/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import styles from "./styles.module.scss";

import React, { useState } from "react";
import { Icon } from "../icon";
import { Menu, MenuItem } from "../menu";
import uniqueId from "lodash/uniqueId";
import { noop } from "lodash";
import { cssNames } from "../../utils";

interface UpdateButtonProps {
  warningLevel?: "light" | "medium" | "high";
  update: () => void;
}

export function UpdateButton({ warningLevel, update }: UpdateButtonProps) {
  const id = uniqueId("update_button_");
  const [opened, setOpened] = useState(false);

  const onKeyDown = (evt: React.KeyboardEvent<HTMLButtonElement>) => {
    if (evt.code == "Space" || evt.code == "Enter") {
      toggle();
    }
  };

  const toggle = () => {
    setOpened(!opened);
  };

  if (!warningLevel) {
    return null;
  }

  return (
    <>
      <button
        data-testid="update-button"
        id={id}
        className={cssNames(styles.updateButton, {
          [styles.warningHigh]: warningLevel === "high",
          [styles.warningMedium]: warningLevel === "medium",
        })}
        onClick={toggle}
        onKeyDown={onKeyDown}
      >
        Update
        <Icon material="update"/>
      </button>
      <Menu
        usePortal
        htmlFor={id}
        isOpen={opened}
        closeOnClickItem
        closeOnClickOutside
        close={toggle}
        open={noop}
      >
        <MenuItem onClick={update} data-testid="update-lens-menu-item">
          Relaunch to Update Lens
        </MenuItem>
      </Menu>
    </>
  );
}
