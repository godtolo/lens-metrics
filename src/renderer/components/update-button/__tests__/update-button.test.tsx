/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { render } from "@testing-library/react";
import React from "react";
import { UpdateButton } from "../update-button";
import "@testing-library/jest-dom/extend-expect";

const update = jest.fn();

describe("<UpdateButton/>", () => {
  beforeEach(() => {
    update.mockClear();
  });

  it("should not render if no warning level prop passed", () => {
    const { queryByTestId } = render(<UpdateButton update={update} />);

    expect(queryByTestId("update-button")).not.toBeInTheDocument();
  });

  it("should render if warning level prop passed", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="light" />);

    expect(getByTestId("update-button")).toMatchSnapshot();
  });

  it("should open menu when clicked", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="light" />);

    const button = getByTestId("update-button");

    button.click();

    expect(getByTestId("update-lens-menu-item")).toBeInTheDocument();
  });

  it("should call update function when menu item clicked", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="light" />);

    const button = getByTestId("update-button");

    button.click();

    const menuItem = getByTestId("update-lens-menu-item");

    menuItem.click();

    expect(update).toHaveBeenCalled();
  });

  it("should have class name with medium warning level", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="medium" />);

    const button = getByTestId("update-button");

    expect(button.className.toLowerCase().includes("medium")).toBeTruthy();
  });

  it("should have class name with high warning level", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="high" />);

    const button = getByTestId("update-button");

    expect(button.className.toLowerCase().includes("high")).toBeTruthy();
  });
});
