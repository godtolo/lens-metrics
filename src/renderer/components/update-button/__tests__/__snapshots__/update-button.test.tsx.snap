// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateButton/> should render if warning level prop passed 1`] = `
<button
  class="updateButton"
  data-testid="update-button"
  id="update_button_2"
>
  Update
  <i
    class="Icon material focusable"
  >
    <span
      class="icon"
      data-icon-name="update"
    >
      update
    </span>
  </i>
</button>
`;
